haystack/preview/components/generators/openai.py

```diff
@@ -0,0 +1,282 @@
+            {
+                "model": chunk.model,
+                "index": 0,
+                "finish_reason": chunk.choices[0].finish_reason,
+                "usage": {},  # we don't have usage data for streaming responses
+            }
+        )
+        return complete_response
+
+    def _build_message(self, completion: OpenAIObject, choice: OpenAIObject) -> ChatMessage:
+        """
+        Converts the response from the OpenAI API to a ChatMessage.
+        :param completion: The completion returned by the OpenAI API.
+        :param choice: The choice returned by the OpenAI API.
+        :return: The ChatMessage.
+        """
+        message: OpenAIObject = choice.message
+        content = dict(message.function_call) if choice.finish_reason == "function_call" else message.content
+        chat_message = ChatMessage.from_assistant(content)
+        chat_message.metadata.update(
+            {
+                "model": completion.model,
+                "index": choice.index,
+                "finish_reason": choice.finish_reason,
+                "usage": dict(completion.usage.items()),
+            }
+        )
+        return chat_message
+
+    def _build_chunk(self, chunk: OpenAIObject, choice: OpenAIObject) -> StreamingChunk:
+        """
+        Converts the response from the OpenAI API to a StreamingChunk.
+        :param chunk: The chunk returned by the OpenAI API.
+        :param choice: The choice returned by the OpenAI API.
+        :return: The StreamingChunk.
+        """
+        has_content = bool(hasattr(choice.delta, "content") and choice.delta.content)
+        if has_content:
+            content = choice.delta.content
+        elif hasattr(choice.delta, "function_call"):
+            content = str(choice.delta.function_call)
+        else:
+            content = ""
+        chunk_message = StreamingChunk(content)
+        chunk_message.metadata.update(
+            {"model": chunk.model, "index": choice.index, "finish_reason": choice.finish_reason}
+        )
+        return chunk_message
+
+    def _check_finish_reason(self, message: ChatMessage) -> None:
+        """
+        Check the `finish_reason` returned with the OpenAI completions.
+        If the `finish_reason` is `length`, log a warning to the user.
+        :param message: The message returned by the LLM.
+        """
+        if message.metadata["finish_reason"] == "length":
+            logger.warning(
+                "The completion for index %s has been truncated before reaching a natural stopping point. "
+                "Increase the max_tokens parameter to allow for longer completions.",
+                message.metadata["index"],
+            )
+        if message.metadata["finish_reason"] == "content_filter":
+            logger.warning(
+                "The completion for index %s has been truncated due to the content filter.", message.metadata["index"]
+            )
+
+    def _post_receive(self, message: ChatMessage) -> None:
```

Comment:
this method only calls `_check_finish_reason`.
why do we need it?

Reply:
We don't. It's a placeholder to add more checks post message receival. Over-engineering.